src/GafferScene/InteractiveRender.cpp

```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+
+class InteractiveRender::SceneGraphOutputFilter : public tbb::thread_bound_filter
+{
+	public:
+	
+		SceneGraphOutputFilter( Renderer *renderer, bool editMode ) :
+			tbb::thread_bound_filter( tbb::filter::serial_in_order ), 
+			m_renderer( renderer ),
+			m_attrBlockCounter( 0 ),
+			m_editMode( editMode )
+		{
+		}
+		
+		~SceneGraphOutputFilter()
+		{
+			// close pending attribute blocks:
+			while( m_attrBlockCounter )
+			{
+				--m_attrBlockCounter;
+				m_renderer->attributeEnd();
+			}
+		}
+		
+		void *operator()( void *item )
+		{
+			SceneGraph *s = (SceneGraph*)item;
+			ScenePlug::ScenePath path;
+			s->path( path );
+			
+			std::string name;
+			ScenePlug::pathToString( path, name );
+
+			try
+			{
+				if( !m_editMode )
+				{
+					// outputting scene for the first time - do some attribute block tracking:
+					if( path.size() )
+					{
+						for( int i = m_previousPath.size(); i >= (int)path.size(); --i )
+						{
+							--m_attrBlockCounter;
+							m_renderer->attributeEnd();
+						}
+					}
+
+					m_previousPath = path;
+
+					++m_attrBlockCounter;
+					m_renderer->attributeBegin();
+
+					// set the name for this location:
+					m_renderer->setAttribute( "name", new StringData( name ) );
+
+				}
+
+				// transform:
+				if( s->m_transform && !m_editMode )
+				{
+					m_renderer->concatTransform( s->m_transform->readable() );
+					s->m_transform = 0;
+				}
+
+				// attributes:
+				if( s->m_attributes )
+				{
+					if( m_editMode )
+					{
+						CompoundDataMap parameters;
+						parameters["exactscopename"] = new StringData( name );
+						m_renderer->editBegin( "attribute", parameters );
+					}
+
+					for( CompoundObject::ObjectMap::const_iterator it = s->m_attributes->members().begin(), eIt = s->m_attributes->members().end(); it != eIt; it++ )
+					{
+						if( const StateRenderable *s = runTimeCast<const StateRenderable>( it->second.get() ) )
+						{
+							s->render( m_renderer );
+						}
+						else if( const ObjectVector *o = runTimeCast<const ObjectVector>( it->second.get() ) )
+						{
+							for( ObjectVector::MemberContainer::const_iterator it = o->members().begin(), eIt = o->members().end(); it != eIt; it++ )
+							{
+								const StateRenderable *s = runTimeCast<const StateRenderable>( it->get() );
+								if( s )
+								{
+									s->render( m_renderer );
+								}
+							}
+						}
+						else if( const Data *d = runTimeCast<const Data>( it->second.get() ) )
+						{
+							m_renderer->setAttribute( it->first, d );
+						}
+					}
+					s->m_attributes = 0;
+
+					if( m_editMode )
+					{
+						m_renderer->editEnd();
+					}
+
+				}
+
+				// object:
+				if( s->m_object && !m_editMode )
+				{
+					if( const VisibleRenderable *renderable = runTimeCast< const VisibleRenderable >( s->m_object.get() ) )
+					{
+						renderable->render( m_renderer );
+					}
+					s->m_object = 0;
+				}
+			}
+			catch( const std::exception& e )
```

Comment:
`&e` not `& e` - could you do a quick search for `& ` everywhere else to make sure there aren't more?